### PR TITLE
User-provided hook within the internal loop of non-linear optimizers

### PR DESCRIPTION
**Pull request by jlblancoc**
See the example in the new unit test. 
This allows progress reports, debugging, GUI updating, etc. during non-linear optimizers are running.

Note: this PR builds upon #585 .

### CODE REVIEW

File: gtsam/nonlinear/NonlinearOptimizerParams.h [hunk 1]
```diff
@@ -33,22 +33,19 @@ namespace gtsam {
...
-  NonlinearOptimizerParams() :
-      maxIterations(100), relativeErrorTol(1e-5), absoluteErrorTol(1e-5), errorTol(
-          0.0), verbosity(SILENT), orderingType(Ordering::COLAMD),
-          linearSolverType(MULTIFRONTAL_CHOLESKY) {}
+  size_t maxIterations = 100; ///< The maximum iterations to stop iterating (default 100)
+  double relativeErrorTol = 1e-5; ///< The maximum relative error decrease to stop iterating (default 1e-5)
+  double absoluteErrorTol = 1e-5; ///< The maximum absolute error decrease to stop iterating (default 1e-5)
+  double errorTol = 0.0; ///< The maximum total error to stop iterating (default 0.0)
+  Verbosity verbosity = SILENT; ///< The printing verbosity during optimization (default SILENT)
+  Ordering::OrderingType orderingType = Ordering::COLAMD; ///< The method of ordering use during variable elimination (default COLAMD)
```
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearOptimizerParams.h:48]: Would it make sense to move the variables to the top and add the constructors after them?

File: gtsam/nonlinear/NonlinearOptimizerParams.h [hunk 2]
```diff
@@ -71,6 +68,15 @@ class GTSAM_EXPORT NonlinearOptimizerParams {
...
   static std::string verbosityTranslator(Verbosity value) ;
 
+  /** Type for an optional user-provided hook to be called after each 
+   * internal optimizer iteration */
+  using IterationHook = std::function<
+    void(size_t /*iteration*/, double/*errorBefore*/, double/*errorAfter*/)>;
+
+  /** Optional user-provided iteration hook to be called after each 
+   * optimization iteration (Default: empty) */
+  IterationHook iterationHook;
```
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearOptimizerParams.h:78]: Can you please elaborate the docs more? One thing is to mention the expected arguments and their types. Perhaps even mention that the iteration hook is a `std::function` object?

**Review by varunagrawal — CHANGES_REQUESTED** [review]
LGTM but requesting some minor updates.

File: gtsam/nonlinear/NonlinearOptimizerParams.h (hunk 1 above)
Comment (jlblancoc) [reply gtsam/nonlinear/NonlinearOptimizerParams.h:48]: Oh, sorry I was actually waiting for feedback on this PR but totally missed your comments! 

> Would it make sense to move the variables to the top and add the constructors after them?

Hmm... it's just a matter of style and consistency I guess. Normally I do it the other way around, but you are right: most other classes in gtsam seems to follow that rule. I'll change it.

File: gtsam/nonlinear/NonlinearOptimizerParams.h (hunk 2 above)
Comment (jlblancoc) [reply gtsam/nonlinear/NonlinearOptimizerParams.h:92]: Done in the latest commit too.

File: gtsam/nonlinear/NonlinearOptimizerParams.h [hunk 3]
```diff
@@ -71,6 +62,37 @@ class GTSAM_EXPORT NonlinearOptimizerParams {
...
+   * context variables, such that the optimizer object itself, the factor graph,
+   * etc.):
+   * \code
+   *  lmOpts.iterationHook = [&](size_t iter, double oldError, double newError)
+   *  {
+   *    // ...
+   *  };
+   * \endcode
+   * or to the result of a properly-formed `std::bind` call.
+   */
```
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearOptimizerParams.h:93]: Amazing!!!

File: gtsam/nonlinear/NonlinearOptimizerParams.h [hunk 4]
```diff
@@ -81,10 +103,16 @@ class GTSAM_EXPORT NonlinearOptimizerParams {
     CHOLMOD, /* Experimental Flag */
   };
 
-  LinearSolverType linearSolverType; ///< The type of linear solver to use in the nonlinear optimizer
+  LinearSolverType linearSolverType = MULTIFRONTAL_CHOLESKY; ///< The type of linear solver to use in the nonlinear optimizer
   boost::optional<Ordering> ordering; ///< The optional variable elimination ordering, or empty to use COLAMD (default: empty)
   IterativeOptimizationParameters::shared_ptr iterativeParams; ///< The container for iterativeOptimization parameters. used in CG Solvers.
 
+  NonlinearOptimizerParams() = default;
```
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearOptimizerParams.h:110]: I remember seeing a comment from @dellaert about having variables before constructors, so he'll be happy about this one. :smile:

**Review by varunagrawal — APPROVED** [review]
LGTM!

**Comment by varunagrawal** [comment]
I believe the build is failing because Actions got rid of Boost 1.69 on Linux virtual environments, opting instead for 1.72. We already landed the PR to upgrade the workflows, but I guess this PR was created before that. I'll push a merge.

**Comment by jlblancoc** [comment]
@dellaert Should we merge this one? Do you want to take a look first?

**Review by dellaert — APPROVED** [review]
Nice! Feel free to merge

**Pull request merged**